src/Umbraco.Infrastructure/PropertyEditors/RichTextEditorPastedImages.cs

```diff
@@ -56,149 +67,301 @@ public sealed class RichTextEditorPastedImages
+            var tmpImgPath = imageTempPath + Path.DirectorySeparatorChar + safeFileName;
+            var absoluteTempImagePath = Path.GetFullPath(tmpImgPath);
 
-                udi = mediaFile.GetUdi();
+            // Convert the base64 content to a byte array and save the bytes directly to a file
+            // this method should work for most use-cases
+            if (encoding.Equals("base64"))
+            {
+                System.IO.File.WriteAllBytes(absoluteTempImagePath, Convert.FromBase64String(imageData));
             }
             else
             {
-                // Already been uploaded & we have it's UDI
-                udi = uploadedImages[tmpImgPath];
+                System.IO.File.WriteAllText(absoluteTempImagePath, HttpUtility.HtmlDecode(imageData), Encoding.UTF8);
             }
 
-            // Add the UDI to the img element as new data attribute
-            img.SetAttributeValue("data-udi", udi.ToString());
+            // When the temp file has been created, we can persist it
+            PersistMediaItem(mediaParentFolder, userId, img, tmpImgPath);
+        }
+
+        return htmlDoc.DocumentNode.OuterHtml;
+    }
+
+    /// <summary>
+    ///     Used by the RTE (and grid RTE) for drag/drop/persisting images.
+    /// </summary>
+    /// <param name="html">HTML from the Rich Text Editor property editor.</param>
+    /// <param name="mediaParentFolder"></param>
+    /// <param name="userId"></param>
+    /// <param name="imageUrlGenerator"></param>
+    /// <returns>Formatted HTML.</returns>
+    [Obsolete("It is not needed to supply the imageUrlGenerator parameter")]
+    public string FindAndPersistPastedTempImages(string html, Guid mediaParentFolder, int userId, IImageUrlGenerator imageUrlGenerator) =>
+        FindAndPersistPastedTempImages(html, mediaParentFolder, userId);
 
-            // Get the new persisted image URL
-            _umbracoContextAccessor.TryGetUmbracoContext(out IUmbracoContext? umbracoContext);
-            IPublishedContent? mediaTyped = umbracoContext?.Media?.GetById(udi.Guid);
-            if (mediaTyped == null)
+    /// <summary>
+    ///     Used by the RTE (and grid RTE) for drag/drop/persisting images.
+    /// </summary>
+    /// <param name="html">HTML from the Rich Text Editor property editor.</param>
+    /// <param name="mediaParentFolder"></param>
+    /// <param name="userId"></param>
+    /// <returns>Formatted HTML.</returns>
+    public string FindAndPersistPastedTempImages(string html, Guid mediaParentFolder, int userId)
+    {
+        // Find all img's that has data-tmpimg attribute
+        // Use HTML Agility Pack - https://html-agility-pack.net
+        var htmlDoc = new HtmlDocument();
+        htmlDoc.LoadHtml(html);
+
+        HtmlNodeCollection? tmpImages = htmlDoc.DocumentNode.SelectNodes($"//img[@{TemporaryImageDataAttribute}]");
+        if (tmpImages == null || tmpImages.Count == 0)
```

Comment:
Lets change this to pattern matching while we're at it 🔥
```suggestion
        if (tmpImages is null || tmpImages.Count == 0)
```